### PR TITLE
Remove argument unpacking from variable kernels

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42043 Remove argument unpacking from variable kernels**
* #42025 IValue stores undefined tensor as None

This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.

Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -201,10 +201,7 @@
 );
 """)
 
-UNPACK_TENSOR = CodeTemplate("""\
-auto${ref} ${arg_name}_ = unpack${suffix}(${arg_name}, "${arg_name}", ${arg_pos});""")
-
-UNPACK_OPTIONS = CodeTemplate("""\
+PACK_TENSOR_OPTIONS = CodeTemplate("""\
```
Comment (bhosmer) [inline tools/autograd/gen_variable_type.py:204]: Why pack? "unpack" is used pretty universally in these scripts to mean "process a param and assign the result to a local", which I think is still what this is doing, right?

**Review by bhosmer — APPROVED** [review]
Nice! Accepting mod a) test failures? and b) inline nit 😁  (also full disclosure, if there's a risk of subtle ripple effects here then mine probably shouldn't be the final review)

**Review by ezyang — APPROVED** [review]
good riddance

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (smessmer) [reply tools/autograd/gen_variable_type.py:204]: that's how "unpack" was used previously but except for tensor options, that is now gone - so afaik, TensorOptions is the only arguments for which we use that nomenclature now. And to "unpack" TensorOptions in my mind means to take a TensorOptions object and scatter it into ScalarType, Layout, Device.

Comment (bhosmer) [reply tools/autograd/gen_variable_type.py:204]: Hmmm what about all this? https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_python_functions.py#L216

**Pull request closed without merging**